crypto/cipher_extra/e_aes_cbc_hmac_sha1.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright 2011-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/opensslconf.h>
+
+#include <stdio.h>
+#include <string.h>
+
+
+#include <openssl/aes.h>
+#include <openssl/cipher.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/objects.h>
+#include <openssl/rand.h>
+#include <openssl/sha.h>
+#include "../fipsmodule/aes/internal.h"
+#include "internal.h"
+
+#if defined(AES_CBC_HMAC_SHA_STITCH)
+
+typedef struct {
+  AES_KEY ks;
+  // Used to compute(init, update and final) HMAC-SHA1.
+  SHA_CTX head, tail, md;
+  // In encrypt case, it's eiv_len + plaintext_len. eiv is explicit iv(required
+  // TLS 1.1+). In decrypt case, it's |EVP_AEAD_TLS1_AAD_LEN(13)|.
+  size_t payload_length;
+  union {
+    unsigned int tls_ver;
+    // In encrypt case, it's not set.
+    // In decrypt case, it stores |additional_data|.
+    // additional_data = seq_num + content_type + protocol_version +
+    // payload_eiv_len seq_num: 8 octets long. content_type: 1 octets long.
+    // protocol_version: 2 octets long.
+    // payload_eiv_len: 2 octets long. eiv is explicit iv required by TLS 1.1+.
+    unsigned char tls_aad[EVP_AEAD_TLS1_AAD_LEN];
+  } aux;
+  // Used after decryption.
+  unsigned char hmac_key[64];
+} EVP_AES_HMAC_SHA1;
+
+#define data(ctx) ((EVP_AES_HMAC_SHA1 *)EVP_CIPHER_CTX_get_cipher_data(ctx))
+
+void aesni_cbc_sha1_enc(const void *inp, void *out, size_t blocks,
+                        const AES_KEY *key, unsigned char iv[16], SHA_CTX *ctx,
+                        const void *in0);
+
+static int aesni_cbc_hmac_sha1_init_key(EVP_CIPHER_CTX *ctx,
+                                        const unsigned char *inkey,
+                                        const unsigned char *iv, int enc) {
+  EVP_AES_HMAC_SHA1 *key = data(ctx);
+  int ret;
+
+  int key_bits = EVP_CIPHER_CTX_key_length(ctx) * 8;
+  if (enc) {
+    ret = aes_hw_set_encrypt_key(inkey, key_bits, &key->ks);
+  } else {
+    ret = aes_hw_set_decrypt_key(inkey, key_bits, &key->ks);
+  }
+
+  SHA1_Init(&key->head);
+  key->tail = key->head;
+  key->md = key->head;
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  return ret < 0 ? 0 : 1;
+}
+
+static int aesni_cbc_hmac_sha1_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                                      const unsigned char *in, size_t len) {
+  EVP_AES_HMAC_SHA1 *key = data(ctx);
+  size_t plen = key->payload_length, iv = 0;
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  if (len % AES_BLOCK_SIZE) {
+    OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_DATA_NOT_MULTIPLE_OF_BLOCK_LENGTH);
+    return 0;
+  }
+
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
```

Comment:
maybe we should have separate functions for encrypt and decrypt because the function is very long and difficult to follow.

Reply:
i don't think separating the encrypt and decrypt help improve any readability here. The two parts are clearly separated with `else` blocks.